File: codeforces/div3_881/d.py
import sys
sys.setrecursionlimit(10**9)
input = lambda: sys.stdin.readline().rstrip()
output = sys.stdout.write

def dfs(node, prevNode):
    global n, graph, leafs

    if leafs[node] != -1:
        return leafs[node]

    if len(graph[node]) == 1 and node != 1:
        leafs[node] = 1
        return 1
    
    l = 0

    for i in graph[node]:
        if i != prevNode:
            l += dfs(i, node)

    leafs[node] = l
    return leafs[node]
    
def solve():
    global n, graph, leafs
    n = int(input())
    graph = [[] for i in range(n + 2)]

    for i in range(n - 1):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)

    leafs = [-1] * (n + 2)

    dfs(1, -1)

    q = int(input())

    for i in range(q):
        qa, qb = map(int, input().split())
        output(f'{leafs[qa] * leafs[qb]}\n')

if __name__ == "__main__":
    for i in range(int(input())):
        solve()

File: codeforces/div3_881/d.java

import java.io.*;
import java.util.*;

public class d{
    public static StringBuilder sb;
    public static BufferedReader br;
    public static String result;
		public static int n;
		public static List<Integer> leafs;
		public static List<List<Integer>> graph;

		public static int dfs(int node, int prevNode)  {
			if(!leafs.get(node).equals(-1)) {
				return leafs.get(node);
			}

			if(graph.get(node).size() == 1 && node != 1) {
				leafs.set(node, 1);
				return leafs.get(node);
			}

			int l = 0;

			for(int nextNode: graph.get(node)) {
				if(nextNode != prevNode) {
					l += dfs(nextNode, node);
				}
			}

			leafs.set(node, l);
			return leafs.get(node);
		}

    public static void solve() throws IOException {
			n = Integer.parseInt(br.readLine());
			graph = new ArrayList<>();

			for(int i = 0; i < n + 1; i++ ) {
				graph.add(new ArrayList<Integer>());
			}

			for(int i = 0; i < n - 1; i++) {
				StringTokenizer st = new StringTokenizer(br.readLine());
				int a = Integer.parseInt(st.nextToken());
				int b = Integer.parseInt(st.nextToken());

				graph.get(a).add(b);
				graph.get(b).add(a);
			}

			leafs = new ArrayList();

			for(int i = 0; i < n + 1; i++) {
				leafs.add(-1);
			}

			dfs(1, -1);

			int q = Integer.parseInt(br.readLine());

			for(int i = 0; i < q; i++) {
				StringTokenizer qst = new StringTokenizer(br.readLine());
				int qa = Integer.parseInt(qst.nextToken());
				int qb = Integer.parseInt(qst.nextToken());

				sb.append((long)leafs.get(qa) * leafs.get(qb)).append("\n");
			}

    }

    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        sb = new StringBuilder();
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
				int t = Integer.parseInt(br.readLine());

				for(int i = 0; i < t; i++) {
					solve();
				}
        result = sb.toString();

        br.close();
        bw.write(result);
        bw.flush();
        bw.close();
    }
}

